docs/DeveloperGuide.md

```diff
@@ -2,37 +2,420 @@
+
+![](.\\uml\\diagrams\\Add.png "AddCommand Class Diagram")
+
+The `AddCommand` component
+- can check if the CS2040CFile to be added into our CLIAlgo exists within the same directory as the program
+- can check for the type of CS2040CFile, whether it is `.txt` or `.cpp` based on the name of the CS2040CFile
+- can ensure that there are no files with repeated names such that all names of files added are unique
+
+## Implementation
+### Filter by keyword feature
+#### Current Implementation
+The filter mechanism is facilitated by `FilterCommand`. It extends `Command` with an overridden `execute()` method. The
+`FilterCommand` has 2 different executions depending on the constructor used to instantiate it. During execution, the
+`FilterCommand` object calls either the `getNotesByTopic()` or the `getAllNotesByTopic()` methods in the `TopicManager`.
+Additionally, it implements the following operations.
+
+- `printAllTopics()` - Prints out all notes stored in CLIAlgo that is sorted by `topic`.
+- `printSingleTopic()` - Prints out all notes stored in CLIAlgo that is tagged to the given `topic`.
+
+The access modifiers of these methods are `private` can can only be accessed within `FilterCommand`.
+
+Given below is an example usage of how the filter mechanism behaves at each step.
+
+> **Step 1**: The user enters a command. The full command is read in by the `Ui` and processed by the `Parser`. If the
+> user entered a valid command, the `Parser` will process the full command using the `StringManipulation` interface and
+> prepare the appropriate `FilterCommand` object.
+
+> **Step 2**: If the `topic` field is left empty, the `Parser` will instantiate a new `FilterCommand` object, setting
+> the `topic` field to be `null`. If the `topic` field is filled with a valid topic name, the `Parser` will instantiate
+> a new `FilterCommand` using its constructor.
+
+### Initializing previous saved data feature
+#### Current implementation
+
+![](.\\sequence\\diagrams\\InitializationFileManager.png "FileManager Initialization Sequence Diagram")
+
+The function for reading the previously saved data is facilitated by the `FileManager`. The `FileManager`
+creates a `SingleFile` for each valid topic name and invokes `createNewFile` for those files in the for 
+in `TOPIC_NAME.txt` in the folder `.\\data`. If the files already exist, they are not created. Instead, 
+the contents of the file would be read line-by-line. The read data would then be passed to `FileDecoder` 
+which would then convert these raw data into `CS2040CFile` objects. The `CS2040File` objects are then passed
+into a `HashMap` which represents the topic these `CS2040CFile` objects belong to. The `HashMap` is then passed
+back to the `TopicManager`, completing the initialization process.
+
+### Export feature
+#### Current implementation
+
+![](.\\sequence\\diagrams\\Export.png "Export Sequence Diagram")
```

Comment:
I think you can explain more about the part of `Parser` to `FilterByTopicCommand` to `TopicManager`, since they are shown on the diagram. Or perhaps it is not that important and can be ignored in the sequence diagram.